subcommands/keys/tuf_utils.go

```diff
@@ -212,44 +212,81 @@ func GetOfflineCreds(credsFile string) (OfflineCreds, error) {
 	return files, nil
 }
 
-func FindTufSigner(keyid, pubkey string, creds OfflineCreds) (*TufSigner, error) {
-	pubkey = strings.TrimSpace(pubkey)
-	for k, v := range creds {
-		if strings.HasSuffix(k, ".pub") {
-			tk := client.AtsKey{}
-			if err := json.Unmarshal(v, &tk); err != nil {
-				return nil, fmt.Errorf("Unable to parse JSON for %s: %w", k, err)
-			}
-			if strings.TrimSpace(tk.KeyValue.Public) == pubkey {
-				pkname := strings.Replace(k, ".pub", ".sec", 1)
-				pkbytes := creds[pkname]
-				tk = client.AtsKey{}
-				if err := json.Unmarshal(pkbytes, &tk); err != nil {
-					return nil, fmt.Errorf("Unable to parse JSON for %s: %w", pkname, err)
-				}
-				keyType, err := parseTufKeyType(tk.KeyType)
-				if err != nil {
-					return nil, fmt.Errorf("Unsupported key type for %s: %s", pkname, tk.KeyType)
-				}
-				pk, err := keyType.ParseKey(tk.KeyValue.Private)
-				if err != nil {
-					return nil, fmt.Errorf("Unable to parse key value for %s: %w", pkname, err)
-				}
-				return &TufSigner{
-					Id:   keyid,
-					Type: keyType,
-					Key:  pk,
-				}, nil
-			}
+func FindOneTufSigner(root *client.AtsTufRoot, creds OfflineCreds, keyids []string) (signer TufSigner, err error) {
+	var signers []TufSigner
+	if signers, err = findTufSigners(root, creds, keyids); err == nil {
+		if len(signers) == 0 {
+			err = fmt.Errorf("Found no active signing key for: %v.", keyids)
+		} else if len(signers) > 1 {
+			err = fmt.Errorf(`Found more than one active signing key for: %v.
+This is an unsupported and insecure way to store private keys.
+Please, provide a keys file which contains a single active signing key.`, keyids)
+		} else {
+			signer = signers[0]
 		}
 	}
-	return nil, fmt.Errorf("Can not find private key for: %s", keyid)
+	return
 }
 
-func findTufRootSigner(root *client.AtsTufRoot, creds OfflineCreds) (*TufSigner, error) {
-	kid := root.Signed.Roles["root"].KeyIDs[0]
-	pub := root.Signed.Keys[kid].KeyValue.Public
-	return FindTufSigner(kid, pub, creds)
+func findTufSigners(root *client.AtsTufRoot, creds OfflineCreds, keyids []string) ([]TufSigner, error) {
+	// Look in creds for each candidate from keyids and return all private keys that match
+	matchPubKeys := make(map[string]client.AtsKey, len(keyids))
+	for _, kid := range keyids {
+		if pk, ok := root.Signed.Keys[kid]; ok {
+			pk.KeyValue.Public = strings.TrimSpace(pk.KeyValue.Public)
```

Comment:
What this trimming is need for? Does it mean that we store pubkey not quite correctly?

Reply:
I don't remember the exact details... IIRC in the past I've seen that public key to be prepended by a newline or smth like that in some edge cases. Note: these key files are not required to be in a canonical form, although all our code tries to generate canonical JSONs. In essence, an RSA key in PEM format with and without a white space is exactly the same. In theory, it can be even split into lines in any way; so, ideally we should split by white space and join without it to compare the keys. But, luckily I never observed a difference in presentations to go that far.

Reply:
My point is that if we store the pub key, then we can store it correctly so the code that reads do not need to do the trimming. Otherwise, it looks like we don't trust ourself.

Anyway, I think the trimming doesn't do any bad.